drivers: i3c: Move clocking code out of NXP I3C driver

mmahadevan108 opened the pull request:
The clocking code present inside the NXP I3C is SoC specific. This has been moved to the SoC folder which makes this driver more flexible to clock architecture variations between SoC's and SDK driver clock API variations.

decsny reviewed (commented):
if these are soc specific, then maybe the binding has been specified wrong. should these clock settings be defined on a separate node representing this clock hardware instead of the i3c node?

dcpleung commented:
Could you separate this into one commit for driver and another for SoC? Also, you need to remove the corresponding devicetree properties in the binding as they are no longer being used.

decsny reviewed (dismissed):
the clock divider properties *must* be removed from the binding for the i3c controller because they are not part of the i3c device, they are part of the rt5xx soc clock tree

decsny commented:
> Could you separate this into one commit for driver and another for SoC? Also, you need to remove the corresponding devicetree properties in the binding as they are no longer being used.

To be pedantic, we need to remove the properties because they don't reflect the reality of hardware, not because they aren't used by the driver anymore. Bindings are not supposed to represent the state of a software project, so if the bindings reflected the actual hardware reality but the driver didn't use them, they wouldn't need to be removed in that case. This way devicetrees could theoretically be used between projects without errors.

But of course this is a completely moot point because:
* there is probably no other project with a binding for this hardware yet (although there may be one showing up in linux soon that we need to coordinate with)
* Zephyr DT currently is not compatible with other projects using DT for a lot of other reasons
* We need to remove the properties for a different reason

mmahadevan108 commented:
> the clock divider properties _must_ be removed from the binding for the i3c controller because they are not part of the i3c device, they are part of the rt5xx soc clock tree

There are other peripheral nodes in the MCXN947 device dts file that follow this model of having clock dividers in the peripheral node. I agree that this clock property belongs in the clock node, however I am not in favor of a mixed implementation where part of the peripheral dividers are in the clock node and part in the peripheral node. Hence for this PR I would prefer to keep the dividers in the peripheral node and address this shift to the clock node as part of a separate cleanup effort.

mmahadevan108 commented:
> Could you separate this into one commit for driver and another for SoC? Also, you need to remove the corresponding devicetree properties in the binding as they are no longer being used.

The device tree binding is getting picked up in the `soc.c` file. I will separate the commits.

mmahadevan108 commented:
> Could you separate this into one commit for driver and another for SoC? Also, you need to remove the corresponding devicetree properties in the binding as they are no longer being used.

Splitting into multiple commits will break bisectability in terms of functionality. I will leave it unchanged.

mmahadevan108 committed:
soc: nxp: Add code to set I3C divider

Update RT5xx and RT6xx clock init to add the code
to set the I3C dividers. This code has been moved
from the I3C driver.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
@decsny @dcpleung, kindly re-review. I have addressed all your comments.

decsny commented on dts/bindings/clock/nxp,lpc-syscon.yaml:
```diff
@@ -14,5 +14,17 @@ properties:
+  i3c-clk-divider-slow:
+    type: array
+    description: Slow clock divider for each I3C instance
+
```
i dont think these belong in the syscon binding either, eventually the clock management framework being proposed should handle this clock as it's own node in the clock tree. Notice there is nothing else in this binding right now related to clock dividers. I dont think we want to go down this route and create more cleanup and have to change bindings in the future.

right now I think best option is what we already do for soc specific system clocking configuration which is just to use soc specific kconfigs.

mmahadevan108 replied:
How about I leave this binding in the `i3c` node and we handle this when the clocking framework is in place instead of adding `Kconfigs` for each SoC we add.

mmahadevan108 committed:
drivers: i3c: Move clocking code out of NXP I3C driver

The clocking code present inside the NXP I3C is SoC specific.
This has been moved to the SoC folder which makes this driver
more flexible to clock architecture variations between SoC's
and SDK driver clock API variations.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
@dcpleung @decsny , can you please revisit

mmahadevan108 commented:
@dcpleung @decsny I have marked the clock divider properties as deprecated. These will be removed in future as part of the clocking rework.

decsny commented:
ci failure addressed https://github.com/zephyrproject-rtos/zephyr/pull/78738

mmahadevan108 commented:
@dcpleung @decsny , kindly revisit this PR
